lib/src/eol.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright 2025 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::error::Error;
+use std::io::Cursor;
+use std::pin::Pin;
+
+use bstr::ByteSlice as _;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncReadExt as _;
+
+use crate::config::ConfigGetError;
+use crate::local_working_copy::TreeStateSettings;
+use crate::settings::UserSettings;
+
+pub(crate) fn create_target_eol_strategy(
+    tree_state_settings: &TreeStateSettings,
+) -> TargetEolStrategy {
+    TargetEolStrategy {
+        eol_conversion_settings: tree_state_settings.eol_conversion_settings,
+    }
+}
+
+fn is_binary(bytes: &[u8]) -> bool {
+    // TODO(06393993): align the algorithm with git so that the git config autocrlf
+    // users won't see different decisions on whether a file is binary and needs to
+    // perform EOL conversion.
+    bytes.contains(&b'\0')
+}
+
+pub(crate) struct TargetEolStrategy {
+    eol_conversion_settings: EolConversionSettings,
+}
+
+impl TargetEolStrategy {
+    /// The limit is to probe whether the file is binary is 8KB.
+    const PROBE_LIMIT: u64 = 8 << 10;
+
+    pub(crate) async fn convert_eol_for_snapshot<'a>(
+        &self,
+        mut contents: impl AsyncRead + Send + Unpin + 'a,
+    ) -> Result<impl AsyncRead + Send + Unpin + 'a, std::io::Error> {
```

Comment:
nit: simply return `Box<dyn ..>` so the caller doesn't have to wrap it in `Box<_>` again.

We don't care much about introducing breaking API changes, and this module is crate-local anyway.

Reply:
Oh, I meant the boxing in `TreeState::write_file()`.

Reply:
Sorry I miss that place. Removed the extra boxing in `TreeState::write_file`. Thanks!